File: acra-core/src/main/java/org/acra/util/StubCreator.java
package org.acra.util;

import android.support.annotation.NonNull;
import org.acra.ACRA;
import org.acra.ErrorReporter;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Proxy;

public final class StubCreator {
    private StubCreator() {
    }

    @NonNull
    public static ErrorReporter createErrorReporterStub() {
        return createStub(ErrorReporter.class, (proxy, method, args) -> {
            String message = ACRA.isACRASenderServiceProcess() ? "in SenderService process" : "before ACRA#init (if you did call #init, check if your configuration is valid)";
            ACRA.log.w(ACRA.LOG_TAG, String.format("ErrorReporter#%s called %s. THIS CALL WILL BE IGNORED!", method.getName(), message));
            return null;
        });
    }

    @NonNull
    public static <T> T createStub(Class<T> interfaceClass, InvocationHandler handler) {
        //noinspection unchecked
        return (T) Proxy.newProxyInstance(StubCreator.class.getClassLoader(), new Class[]{interfaceClass}, handler);
    }
}

File: acra-core/src/main/java/org/acra/util/StubCreator.kt
package org.acra.util

import org.acra.ACRA.isACRASenderServiceProcess
import org.acra.ErrorReporter
import org.acra.log.warn
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Proxy

object StubCreator {
    fun createErrorReporterStub(): ErrorReporter {
        return createStub { _, method, _ ->
            val message = if (isACRASenderServiceProcess()) "in SenderService process" else "before ACRA#init (if you did call #init, check if your configuration is valid)"
            warn { "ErrorReporter#${method.name} called $message. THIS CALL WILL BE IGNORED!" }
            null
        }
    }

    inline fun <reified T> createStub(handler: InvocationHandler): T = createStub(T::class.java, handler)

    @JvmStatic
    fun <T> createStub(interfaceClass: Class<T>, handler: InvocationHandler): T {
        @Suppress("UNCHECKED_CAST")
        return Proxy.newProxyInstance(StubCreator::class.java.classLoader, arrayOf<Class<*>>(interfaceClass), handler) as T
    }
}